src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java

```diff
@@ -369,39 +452,57 @@ public CompletableFuture<Mqtt5Client> connect() {
-                            return UnsubscribeResponse.fromCrtUnsubAck(r);
-                        });
-            });
+            return connect()
+                    .thenCompose(
+                            (client) -> {
+                                logger.atDebug()
+                                        .kv(TOPIC_KEY, topic)
+                                        .log("Unsubscribing from topic");
+                                return client.unsubscribe(
+                                                new UnsubscribePacket.UnsubscribePacketBuilder()
+                                                        .withSubscription(topic)
+                                                        .build())
+                                        .thenApply(
+                                                r -> {
+                                                    try (LockScope ls2 = LockScope.lock(lock)) {
+                                                        subscriptionTopics.removeIf(
+                                                                s -> s.getTopic().equals(topic));
+                                                    }
+                                                    return UnsubscribeResponse.fromCrtUnsubAck(r);
+                                                });
+                            });
         }
     }
 
     @Override
     public CompletableFuture<PubAck> publish(Publish publish) {
         try (LockScope ls = LockScope.lock(lock)) {
-            return connect().thenCompose((client) -> {
-                // Take the tokens from the limiters' token buckets.
-                // This is guaranteed to not block because we've already slept the required time
-                // in the spooler thread before calling this method.
-                transactionLimiter.acquire();
-                bandwidthLimiter.acquire(publish.getPayload().length);
-                logger.atTrace().kv(TOPIC_KEY, publish.getTopic()).kv(QOS_KEY, publish.getQos().name())
-                        .log("Publishing message");
-                return client.publish(publish.toCrtPublishPacket()).thenApply(r -> {
-                    if (r.getType().equals(PublishResult.PublishResultType.NONE)) {
-                        return new PubAck(0, null, null);
-                    }
-                    PubAckPacket p = r.getResultPubAck();
-                    return PubAck.fromCrtPubAck(p);
-                });
-            });
+            return connect()
+                    .thenCompose(
+                            (client) -> {
+                                // Take the tokens from the limiters' token buckets.
+                                // This is guaranteed to not block because we've already slept the
+                                // required time
+                                // in the spooler thread before calling this method.
+                                transactionLimiter.acquire();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Lock acquisition without timeout detected. Using locks without timeouts can lead to  indefinite blocking and service outages if the lock cannot be acquired. Use timeout-based methods like tryLock(timeout), tryAcquire(timeout), or await(timeout) instead. Learn more on : https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Lock.html#tryLock-long-java.util.concurrent.TimeUnit